python/tvm/autotvm/tophub.py

```diff
@@ -110,6 +110,10 @@ def context(target, extra_files=None):
         device = tgt.attrs.get("device", "")
         if device != "":
             possible_names.append(_alias(device))
+        # for cases when we do have explicitly defined -device in the target,
+        # we still might have information about it stored in keys container
+        # in other case we will load statistics for definitely irrelative stat
```

Comment:
Any chance this comment can be made clearer? :) I read it several times and I don't get it. 
`might have information`
What information?
`about it`
About what?
`keys container`
I guess that is Target.keys? Maybe make that more explicit
`in other case`
What case? When keys container is empty or when -device is not defined?
`irrelative stat`
What is irrelative stat?

Also, is it one or multiple sentences? Capital letters and punctuation would be helpful :)

Reply:
:) tried to include logic of matching statistics without detailed description in this place.

1. the only mandatory field in target is `kind`. it can be llvm/opencl/cuda/...
2. it might be `device` specified, but not required. It might be "arm_cpu", or "adreno" for example
3. it might be `model` specified, but not required. It might be "riscv32-unknown-linux-gnu" or "kirin970" for example
4. the `keys` field is not specified but is filled after analysis of the Target string automatically by devices. It is done for hierarchical search in the `topi` implementations and do fallback to lower/parents layers

Before this change, dedicated tophub *file* was searched only by `device` and second by `kind` 
As you remember, `device` is not mandatory in the Target. At the same time TVM understands that target belongs to arm_cpu, it treat it as arm_cpu, but before my change, the target like `llvm -model=snapdragon835 -mtriple=arm64-linux-android -mattr=+neon` was treated only as `llvm` stat file. I cannot even imagine on which device `llvm` stat was collected, but definitely not on ARM. I.e. statistics was irrelevant.

Adding keys to the priority before `kind` we increase the probability of searching proper statistics.

Do I need to comment line by line your answer?

Reply:
ok, will modify comment

Reply:
As for applying of the statistics, I am not even sure that taking statistics from another model of the same device is a good idea. Fortunately it more or less works for ARM CPU currently but if we take intel graphics, the previous models of Iris had different limitation of group worksize. Statistics which collected on newer graphics cannot be applied on previous one. I.e. we risk not to have bad performance but have just not working kernels.

I.e. I agree that we need to do something with tophub and search algorithm, but do not have idea how we need to do this.

And probably it is not so important assuming that Metaschedule should be the tuner #1. On the other hand MS is not a solution for default compilation without tuning.

Reply:
Yeah I agree that applying same tuning statistics across different implementations of the same architecture is quite sketchy, especially if it happens in this non-transparent way it currently does. I'm also not sure how often the database gets updated with new stats. I haven't seen compilation crashing because of unsuitable tuning records from tophub myself, but if it happens, that's really bad, since it makes tuning compulsory for being able to compile a network with TVM. Also, sounds like a really difficult thing to debug. But I suppose as long as AutoTVM is the default tuner (as it is for tvmc) and therefore the one new potential users will use to evaluate TVM, it should be kept in a state where its behaviour is representative of TVM's capabilities. In my opinion it would be good if we didn't have 3 tuners to maintain in parallel, but I suppose that's where the community is at the moment.